ingest.py

```diff
@@ -1,28 +1,149 @@
+from langchain.document_loaders.parsers import LanguageParser
+from langchain.storage import RedisStore, EncoderBackedStore
+from langchain.retrievers import ParentDocumentRetriever
+from langchain.utilities.redis import get_client
+import json
 
+WEAVIATE_URL=os.environ["WEAVIATE_URL"]
+WEAVIATE_API_KEY=os.environ["WEAVIATE_API_KEY"]
 
+def ingest_repo():
+    repo_path = os.path.join(os.getcwd(), "test_repo")
+    if os.path.exists(repo_path):
+        shutil.rmtree(repo_path)
+
+    repo = Repo.clone_from("https://github.com/langchain-ai/langchain", to_path=repo_path)
+
+    loader = GenericLoader.from_filesystem(
+        repo_path+"/libs/langchain/langchain",
+        glob="**/*",
+        suffixes=[".py"],
+        parser=LanguageParser(language=Language.PYTHON, parser_threshold=500)
+    )
+    documents_repo = loader.load()
+    len(documents_repo)
+
+    python_splitter = RecursiveCharacterTextSplitter.from_language(language=Language.PYTHON, 
+                                                                chunk_size=2000, 
+                                                                chunk_overlap=200)
+    texts = python_splitter.split_documents(documents_repo)
+    return texts
+
+def key_encoder(key: int) -> str:
+    return json.dumps(key)
+
+def value_serializer(value: float) -> str:
+    if isinstance(value, Document):
+        value = {
+            'page_content': value.page_content,
+            'metadata': value.metadata,
+        }
+    return json.dumps(value)
+
+def value_deserializer(serialized_value: str) -> Document:
+    value = json.loads(serialized_value)
+    if 'page_content' in value and 'metadata' in value:
+        return Document(page_content=value['page_content'], metadata=value['metadata'])
+    else:
+        return value
+    
 def ingest_docs():
     """Get documents from web pages."""
-    loader = ReadTheDocsLoader("langchain.readthedocs.io/en/latest/")
-    raw_documents = loader.load()
-    text_splitter = RecursiveCharacterTextSplitter(
-        chunk_size=1000,
-        chunk_overlap=200,
+
+    urls = [
+        "https://api.python.langchain.com/en/latest/api_reference.html#module-langchain",
+        "https://python.langchain.com/docs/get_started", 
+        "https://python.langchain.com/docs/use_cases",
+        "https://python.langchain.com/docs/integrations",
+        "https://python.langchain.com/docs/modules", 
+        "https://python.langchain.com/docs/guides",
+        "https://python.langchain.com/docs/ecosystem",
+        "https://python.langchain.com/docs/additional_resources",
+        "https://python.langchain.com/docs/community",
+    ]
+
+    documents = []
+    for url in urls:
+        loader = RecursiveUrlLoader(url=url, max_depth=2 if url == urls[0] else 10, extractor=lambda x: Soup(x, "lxml").text, prevent_outside=True)
+        temp_docs = loader.load()
+        temp_docs = [doc for i, doc in enumerate(temp_docs) if doc not in temp_docs[:i]]        
```

Comment:
this is probably not very efficient and also is slightly confusing. coud just set `set(temp_docs)`?

Reply:
also why is this necessary? are we scraping the same page twice?

Reply:
true, deleted